app/services/foreman_bootdisk/iso_generator.rb

```diff
@@ -85,8 +85,13 @@ def self.generate(opts = {})
         EOT
       end
 
-      # Temporary directory cannot be cleaned inprocess due to send_file offloaded to web server
-      wd = Dir.mktmpdir('bootdisk-iso-')
+      # Temporary directory cannot be cleaned in-process due to send_file offloaded to web server and
+      # the call is asynchronous. Also $TMPDIR (/tmp or /var/tmp) cannot be used because Foreman runs
+      # with PrivateTmp systemd setting. And the plugin does not ship its own systemd-tmpfiles config,
```

Comment:
I don't understand this reasoning. Why can't `$TMPDIR` be used because it's on `PrivateTmp`? To me it doesn't appear related at all. On Debian you have pretty much the same guarantees for `Rails.root.join('tmp')` as for `$TMPDIR`: it only exists while the service is running and is destroyed on service restart. On Red Hat we do place a tmpfiles entry so it is persisted between runs.

Also, offloading `send_file` to the web server probably doesn't work since both `$TMPDIR` and `Rails.root.join('tmp')` live in directories that only the Foreman user can read so I think sendfile support doesn't work. I also don't think sendfile works with a reverse proxy setup.

So really, I think the commend should be "this allows for an easy cherry pick"

Reply:
> Why can't $TMPDIR be used because it's on PrivateTmp?

That is just a bad wording. This is what I mean:

```
        # Temporary directory cannot be cleaned in-process due to asynchronous send_file call.
        # Also we cannot rely on systemd-tmpfiles-clean as private temporary files are not subject
        # of scheduled cleanups. Let's clean bootdisks from prevous requests manually.
```

> Also, offloading send_file to the web server probably doesn't work

But it works. Frankly, I gave it about 15 minutes or research and I don't understand how this is possible - there is not a single mention of send_file in Puma, Rails appears to call `send_file` from `response` (Rack object?) and in Rack documentation there are few words about selected http servers: https://www.rubydoc.info/gems/rack/Rack/Sendfile and it appears they need to be configured in a special way which we do not do I suppose.

But it works, there must be some kind of fallback mechanism. I just tried it on a 2.4 RC3 production install to be sure.

> So really, I think the commend should be "this allows for an easy cherry pick"

Well, this is an unfair comment. I have volunteered to maintain the repository but I never had ambitions to refactor it in a big way. I am trying to keep this plugin alive.

There are two options: `send_file` which is known to work in production for years, and `send_data` which would allow synchronous call (thus cleanup after the download) but it would require the Rails process to load the whole file into memory so it is probably harder on memory. There is also an option to stream the data manually via another Rails API but since I am already having issues in the other (core) PR with that, I am really hesitant to start anything like this here.

Reply:
> But it works, there must be some kind of fallback mechanism. I just tried it on a 2.4 RC3 production install to be sure.

I'd think it falls back to sending the file the old fashioned way: send bytes from a Ruby process.